lib/truffle/socket/tcp_server.rb

```diff
@@ -25,24 +25,29 @@
 # EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 class TCPServer < TCPSocket
-  def initialize(host, service = nil)
+  NOT_PASSED = Object.new
+
+  def initialize(host, service = NOT_PASSED)
     @no_reverse_lookup = self.class.do_not_reverse_lookup
 
-    if host.is_a?(Integer) and service.nil?
-      service = host
-      host    = nil
-    end
-
-    if host.is_a?(String) and service.nil?
-      begin
-        service = Integer(host)
-      rescue ArgumentError
-        raise SocketError, "invalid port number: #{host}"
-      end
+    if NOT_PASSED.equal?(service)
+      service =
+        if host.is_a?(Integer)
+          host
+        else
+          string_service = Truffle::Socket.coerce_to_string(host)
+          begin
+            Integer(string_service)
+          rescue ArgumentError
+            raise SocketError, "invalid port number: #{host}"
+          end
+        end
 
       host = nil
     end
```

Comment:
Do you think we could model this more simply by using:
```ruby
def initialize(host = nil, service)
```
It seems that if we are using just 1 argument, it's always for the service and host being `nil`/`''`.

Reply:
Thanks! I didn't think of using a trailing required arg. Actually, it seems that using
```
def initialize(host = nil, service)
```
we can simply delegate all the arguments to `Socket` like:
```diff
diff --git a/lib/truffle/socket/tcp_server.rb b/lib/truffle/socket/tcp_server.rb
index dcbb7dbeff..76e6e86197 100644
--- a/lib/truffle/socket/tcp_server.rb
+++ b/lib/truffle/socket/tcp_server.rb
@@ -27,37 +27,9 @@
 class TCPServer < TCPSocket
   NOT_PASSED = Object.new

-  def initialize(host, service = NOT_PASSED)
+  def initialize(host = nil, service)
     @no_reverse_lookup = self.class.do_not_reverse_lookup

-    if NOT_PASSED.equal?(service)
-      service =
-        if host.is_a?(Integer)
-          host
-        else
-          string_service = Truffle::Socket.coerce_to_string(host)
-          begin
-            Integer(string_service)
-          rescue ArgumentError
-            raise SocketError, "invalid port number: #{host}"
-          end
-        end
-
-      host = nil
-    end
-
-    service = 0 if service.nil?
-
-    unless service.is_a?(Integer)
-      service = Truffle::Socket.coerce_to_string(service)
-    end
-
-    if host
-      host = Truffle::Socket.coerce_to_string(host)
-    else
-      host = ''
-    end
-
     remote_addrs = Socket
       .getaddrinfo(host, service, :UNSPEC, :STREAM, 0, Socket::AI_PASSIVE)
```
I still need to test this a bit more but does this sound better?

Reply:
Yes, that sounds great, could you update the PR with that?

Reply:
Updated.